File: ios/SkyCast/Core/DesignSystem/WeatherBackground.swift
import SwiftUI

/// How strongly a weather background reads.
///
/// Today gets the full treatment because it is the screen about the weather right now. Everything
/// else gets a whisper of the same hue: enough that the app feels like one place, not so much that
/// a list of saved cities competes with the forecast for attention.
enum BackgroundIntensity {
    case full
    case subtle

    var topOpacity: Double {
        switch self {
        case .full: 0.30
        case .subtle: 0.14
        }
    }

    var midOpacity: Double {
        switch self {
        case .full: 0.10
        case .subtle: 0.05
        }
    }

    var glowOpacity: Double {
        switch self {
        case .full: 0.22
        case .subtle: 0.10
        }
    }

    /// What the wash settles to at the very bottom of the screen, rather than fading to nothing.
    var floorOpacity: Double {
        switch self {
        case .full: 0.06
        case .subtle: 0.03
        }
    }
}

/// A background that reflects the current condition and time of day.
///
/// Paints the system grouped background and lays a **low-opacity wash of the condition's hue** over
/// it, so the base is always the colour the system guarantees text contrast against and the
/// condition shifts the mood rather than replacing the palette.
///
/// The colour also gives Liquid Glass something to sample: a flat white page renders glass as a
/// faint grey rectangle.
///
/// The wash drifts slowly, at roughly half a minute per cycle, and stops entirely under **Reduce
/// Motion**.
///
/// The Android counterpart is `core/designsystem/component/WeatherBackground.kt`.
struct WeatherBackground: View {
    let condition: WeatherCondition
    let isDaytime: Bool
    var intensity: BackgroundIntensity = .full

    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @State private var drift: Double = WeatherBackground.driftMidpoint

    private var tint: Color {
        WeatherPalette.tint(for: condition, isDaytime: isDaytime)
    }

    var body: some View {
        GeometryReader { proxy in
            ZStack {
                Color.skyBackground

                // Runs the full height, not to `.center`, so there is no hard edge across the page
                // and the area behind the glass tab bar is not flat background. The floor opacity is
                // non-zero for the same reason: glass needs something behind it to sample.
                LinearGradient(
                    stops: [
                        .init(color: tint.opacity(intensity.topOpacity), location: 0),
                        .init(color: tint.opacity(intensity.midOpacity), location: 0.45),
                        .init(color: tint.opacity(intensity.floorOpacity), location: 1),
                    ],
                    startPoint: .top,
                    endPoint: .bottom
                )

                // A soft glow standing in for where the light is coming from. Both its position
                // and its strength move: position alone was imperceptible, changing any given
                // pixel by about one value in 255, which is not an animation but a rounding error.
                RadialGradient(
                    colors: [
                        tint.opacity(intensity.glowOpacity * (Self.glowOpacityMin + drift * Self.glowOpacityTravel)),
                        .clear,
                    ],
                    center: UnitPoint(
                        x: Self.glowXMin + drift * Self.glowXTravel,
                        y: Self.glowY
                    ),
                    startRadius: 0,
                    endRadius: proxy.size.width * Self.glowRadiusRatio
                )
            }
        }
        .ignoresSafeArea()
        // Decorative by definition: it carries mood, and every fact it hints at is stated in text
        // elsewhere on the screen.
        .accessibilityHidden(true)
        .onAppear(perform: startDrift)
        .onChange(of: reduceMotion) { _, _ in startDrift() }
    }

    private func startDrift() {
        guard !reduceMotion else {
            drift = Self.driftMidpoint
            return
        }
        withAnimation(.linear(duration: Self.driftDuration).repeatForever(autoreverses: true)) {
            drift = 1
        }
    }

    /// Slow enough to read as changing light rather than as something moving.
    private static let driftDuration: TimeInterval = 18
    private static let driftMidpoint = 0.5
    private static let glowXMin = 0.15
    private static let glowXTravel = 0.7
    private static let glowY = 0.1

    /// Tighter than the full width, so the glow reads as a source of light rather than a haze.
    private static let glowRadiusRatio = 0.6

    /// The glow breathes between 70% and 130% of its base strength as it travels.
    private static let glowOpacityMin = 0.7
    private static let glowOpacityTravel = 0.6
}

extension View {
    /// Places a weather background behind this view.
    func weatherBackground(
        condition: WeatherCondition,
        isDaytime: Bool,
        intensity: BackgroundIntensity = .full
    )
        -> some View
    {
        background {
            WeatherBackground(condition: condition, isDaytime: isDaytime, intensity: intensity)
        }
        // Published to the content in front, not just painted behind it: the detail tiles and any
        // other surface on the page mix a little of it into their own fill.
        .environment(\.weatherTint, WeatherPalette.tint(for: condition, isDaytime: isDaytime))
    }
}

File: android/core/designsystem/src/main/kotlin/com/nauhaan/skycast/core/designsystem/component/WeatherBackground.kt
package com.nauhaan.skycast.core.designsystem.component

import android.provider.Settings
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import com.nauhaan.skycast.core.designsystem.theme.LocalWeatherTint
import com.nauhaan.skycast.core.designsystem.theme.weatherTint
import com.nauhaan.skycast.domain.model.WeatherCondition

/**
 * A background that reflects the current condition and time of day.
 *
 * Paints the theme's own `surface` and lays a **low-alpha wash of the condition's hue** over it, so
 * the base is always the colour the theme guarantees text contrast against and the condition shifts
 * the mood rather than replacing the palette.
 *
 * The wash drifts slowly, at eighteen seconds per sweep, and stops entirely when the user has asked
 * for less motion.
 */
@Composable
fun WeatherBackground(
    condition: WeatherCondition,
    isDaytime: Boolean,
    modifier: Modifier = Modifier,
    intensity: BackgroundIntensity = BackgroundIntensity.FULL,
    content: @Composable BoxScope.() -> Unit,
) {
    val surface = MaterialTheme.colorScheme.surface
    val tint = weatherTint(condition, isDaytime)
    val drift = animatedDrift()

    Box(
        modifier = modifier
            .fillMaxSize()
            .drawBehind {
                drawRect(surface)

                // Two washes: a broad vertical one for the overall mood, and a soft radial glow
                // standing in for where the light is coming from. The radial one is what moves.
                drawRect(
                    brush = Brush.verticalGradient(
                        colors = listOf(
                            tint.copy(alpha = intensity.topAlpha),
                            tint.copy(alpha = intensity.midAlpha),
                            // Settles to a trace, so the bottom of the screen, where the navigation
                            // bar sits, still belongs to the page.
                            tint.copy(alpha = intensity.floorAlpha),
                        ),
                    ),
                )
                // Both the position and the strength move. Position alone was imperceptible:
                // a wide, soft glow sliding across a wide, soft gradient changes any given pixel
                // by about one value in 255, which is not an animation, it is a rounding error.
                val glowAlpha = intensity.glowAlpha * (GLOW_ALPHA_MIN + drift * GLOW_ALPHA_TRAVEL)
                drawRect(
                    brush = Brush.radialGradient(
                        colors = listOf(tint.copy(alpha = glowAlpha), Color.Transparent),
                        center = Offset(
                            x = size.width * (GLOW_X_MIN + drift * GLOW_X_TRAVEL),
                            y = size.height * GLOW_Y,
                        ),
                        radius = size.width * GLOW_RADIUS_RATIO,
                    ),
                )
            },
    ) {
        // Published to the content in front, not just painted behind it: the detail tiles and any
        // other surface on the page mix a little of it into their own fill.
        CompositionLocalProvider(LocalWeatherTint provides tint) {
            content()
        }
    }
}

/**
 * How strongly the background reads.
 *
 * Today gets the full treatment because it is the screen about the weather right now. Everything
 * else gets a whisper of the same hue: enough that the app feels like one place, not so much that a
 * list of saved cities competes with the forecast for attention.
 */
enum class BackgroundIntensity(
    val topAlpha: Float,
    val midAlpha: Float,
    val glowAlpha: Float,
    /** What the wash settles to at the very bottom of the screen. */
    val floorAlpha: Float,
) {
    FULL(topAlpha = 0.30f, midAlpha = 0.10f, glowAlpha = 0.22f, floorAlpha = 0.06f),
    SUBTLE(topAlpha = 0.14f, midAlpha = 0.05f, glowAlpha = 0.10f, floorAlpha = 0.03f),
}

/**
 * A 0–1 value that drifts back and forth, or a fixed midpoint when motion is reduced.
 *
 * Checked through `ANIMATOR_DURATION_SCALE`, which is what "Remove animations" in Android's
 * accessibility settings actually sets, and what the platform's own animators consult.
 */
@Composable
private fun animatedDrift(): Float {
    val context = LocalContext.current
    val animationsEnabled = remember(context) {
        Settings.Global.getFloat(
            context.contentResolver,
            Settings.Global.ANIMATOR_DURATION_SCALE,
            1f,
        ) != 0f
    }

    if (!animationsEnabled) return DRIFT_MIDPOINT

    val transition = rememberInfiniteTransition(label = "weatherBackground")
    val drift by transition.animateFloat(
        initialValue = 0f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(
            // Linear and slow. An eased or bouncing background would draw the eye, which is the
            // opposite of what a background is for.
            animation = tween(durationMillis = DRIFT_DURATION_MILLIS, easing = LinearEasing),
            repeatMode = RepeatMode.Reverse,
        ),
        label = "weatherBackgroundDrift",
    )
    return drift
}

/** Slow enough to read as changing light. */
private const val DRIFT_DURATION_MILLIS = 18_000
private const val DRIFT_MIDPOINT = 0.5f
private const val GLOW_X_MIN = 0.15f
private const val GLOW_X_TRAVEL = 0.7f
private const val GLOW_Y = 0.1f

/** Tighter than the full width, so the glow reads as a source of light. */
private const val GLOW_RADIUS_RATIO = 0.6f

/** The glow breathes between 70% and 130% of its base strength as it travels. */
private const val GLOW_ALPHA_MIN = 0.7f
private const val GLOW_ALPHA_TRAVEL = 0.6f
